### PR TITLE
Upload errors.pb in new API

### PR DESCRIPTION
**Pull request by wild-endeavor**
# TL;DR
Currently, new style Python-native typed tasks that raise an exception do not properly write an errors.pb file.  This replicates the existing logic in the flyte engine task code.

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
Copied code from https://github.com/flyteorg/flytekit/blob/master/flytekit/engines/flyte/engine.py#L320-L376

## Tracking Issue
https://github.com/flyteorg/flyte/issues/739

### CODE REVIEW

File: flytekit/bin/entrypoint.py [hunk 1]
```diff
@@ -93,23 +98,40 @@ def _dispatch_execute(ctx: FlyteContext, task_def: PythonTask, inputs_path: str,
...
         if isinstance(e, IgnoreOutputs):
             # Step 3b
             _logging.warning(f"IgnoreOutputs received! Outputs.pb will not be uploaded. reason {e}")
             return
         # Step 3c
         _logging.error(f"Exception when executing task {task_def.name}, reason {str(e)}")
-        raise e
+        _logging.error("!! Begin Unknown System Error Captured by Flyte !!")
+        exc_str = _traceback.format_exc()
+        output_file_dict[_constants.ERROR_FILE_NAME] = _error_models.ErrorDocument(
```
Comment (kumare3) [inline flytekit/bin/entrypoint.py:124]: So I wanted to do this - in case of error (and this maybe because we are unable to write to S3/GCS etc) we should write the output to https://kubernetes.io/docs/tasks/debug-application-cluster/determine-reason-pod-failure/#customizing-the-termination-message
This would make propeller read the outputs and write it to Flyteadmin.

Comment (wild-endeavor) [reply flytekit/bin/entrypoint.py:124]: Can we backlog this as it's a bit different?

Comment (EngHabu) [reply flytekit/bin/entrypoint.py:124]: We kinda do that... [Here](https://github.com/flyteorg/flyteplugins/blob/8a49ee2bbef9e89b6d1cec94b4eef1ee8fe7aca3/go/tasks/pluginmachinery/flytek8s/container_helper.go#L127)
and we do indeed take that message and bubble up to Admin (after shortening it)

File: flytekit/bin/entrypoint.py [hunk 2]
```diff
@@ -93,23 +98,40 @@ def _dispatch_execute(ctx: FlyteContext, task_def: PythonTask, inputs_path: str,
...
+            _error_models.ContainerError("SYSTEM:Unknown", exc_str, _error_models.ContainerError.Kind.RECOVERABLE,)
+        )
+        _logging.error(exc_str)
+        _logging.error("!! End Error Captured by Flyte !!")
+
+    for k, v in output_file_dict.items():
+        _common_utils.write_proto_to_file(v.to_flyte_idl(), _os.path.join(ctx.execution_state.engine_dir, k))
+
+    ctx.file_access.upload_directory(ctx.execution_state.engine_dir, output_prefix)
```
Comment (EngHabu) [inline flytekit/bin/entrypoint.py:133]: So what happens if this fails?

Comment (EngHabu) [reply flytekit/bin/entrypoint.py:133]: This is the issue we talked about earlier, isn't it? anyway we can capture that and write the original error (+ the new) to stdErr?

File: flytekit/bin/entrypoint.py (hunk 1 above)
Comment (kumare3) [reply flytekit/bin/entrypoint.py:124]: @EngHabu I know we do that in FlytePropeller, what I meant was do this in flytekit as well?

Comment (EngHabu) [reply flytekit/bin/entrypoint.py:124]: Hmmm I'm not sure what do we NOT do currently (after this PR)? in the older flytekit, when we fail to write the outputs, we write THAT error into stdErr and that bubbles up... are you saying with this PR we don't yet do that? or are you saying "Also write the real outputs" to stdErr?

File: flytekit/bin/entrypoint.py (hunk 2 above)
Comment (wild-endeavor) [reply flytekit/bin/entrypoint.py:133]: The exception gets caught at a higher layer and you see the error message you don't like https://nuclyde.slack.com/archives/D01FD1XGYRZ/p1612578571113500

File: flytekit/bin/entrypoint.py (hunk 1 above)
Comment (kumare3) [reply flytekit/bin/entrypoint.py:124]: no, i do not follow, where we write the error to stdErr

**Review by kumare3 — APPROVED** [review]
